rainforest/mobile_upload.go

```diff
@@ -0,0 +1,212 @@
+func (c *Client) UploadToS3(postData *RFPresignedPostData, filePath string) error {
+	var req *http.Request
+	fileName := filepath.Base(filePath)
+
+	f, err := os.Open(filePath)
+	defer f.Close()
+
+	fi, err := f.Stat()
+	if err != nil {
+		return err
+	}
+	fileSize := int64(fi.Size())
+
+	contentLength := postData.emptyMultipartSize("file", fileName) + fileSize
+
+	readBody, writeBody := io.Pipe()
+	defer readBody.Close()
+
+	writer := multipart.NewWriter(writeBody)
+
+	// Do the writes async streamed
+	errChan := make(chan error, 1)
+	go func() {
+		defer writeBody.Close()
+
+		// Add the required fields from S3
+		addFields(postData.RequiredFields, writer)
+
+		part, err := writer.CreateFormFile("file", fileName)
+		if err != nil {
+			errChan <- err
+			return
+		}
+		if _, err := io.CopyN(part, f, fileSize); err != nil {
+			errChan <- err
+			return
+		}
+		errChan <- writer.Close()
+	}()
+
+	// Create the Request
+	url := postData.URL
+	req, err = http.NewRequest("POST", url, readBody)
+	if err != nil {
+		<-errChan
+		return err
+	}
+	req.Header.Set("Content-Type", writer.FormDataContentType())
+	req.ContentLength = contentLength
+
+	// Perform the upload
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		<-errChan
+		return err
+	}
+
+	status := resp.StatusCode
+	if status >= 300 {
+		body, err := ioutil.ReadAll(resp.Body)
+
+		<-errChan
+		if err != nil {
+			return err
+		}
+
+		return fmt.Errorf("There was an error uploading your file - %v: %v", fileName, string(body))
+	}
+
+	return <-errChan
+}
+
+func addFields(requiredFields map[string]string, writer *multipart.Writer) {
+	for k, v := range requiredFields {
+		writer.WriteField(k, v)
+	}
+}
+
+func (postData *RFPresignedPostData) emptyMultipartSize(fieldname, filename string) int64 {
+	body := &bytes.Buffer{}
+	writer := multipart.NewWriter(body)
+	addFields(postData.RequiredFields, writer)
+	writer.CreateFormFile("file", filename)
+	writer.Close()
+	return int64(body.Len())
+}
+
+// UpdateURL fetches sites available to use during the RF runs.
+func (c *Client) UpdateURL(siteID int, environmentID int, newURL string) error {
+	siteEnvironmentID, err := c.getSiteEnvironments(siteID, environmentID)
+	if err != nil {
+		return err
+	}
+
+	err = c.setSiteEnvironmentURL(siteEnvironmentID, newURL)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// SiteEnvironmentsData type represents the collection returned when calling the site_environments get endpoint
+type SiteEnvironmentsData struct {
+	SiteEnvironments []SiteEnvironment `json:"site_environments"`
+}
+
+// SiteEnvironment type represents a single SiteEnvironment returned by the API call for a list of sites
+type SiteEnvironment struct {
+	ID            int    `json:"id"`
+	SiteID        int    `json:"site_id"`
+	EnvironmentID int    `json:"environment_id"`
+	URL           string `json:"url"`
+}
+
+func (c *Client) getSiteEnvironments(siteID int, environmentID int) (int, error) {
```

Comment:
This should be `getSiteEnvironmentID` or something right? It's only returning the ID. Also do we normally return -1 or 0 alongside errors?

Reply:
I see we return 0 in a few other places so switched it to that.